File: lib/std/core/types/timer.lua
require 'std.core.util'
require 'std.core.game'
local hook = require 'std.core.hook'

timer = timer or {}

local context = util.context("timer", {
    scheduled = {}
})

local scheduledTimers = context.scheduled
local timerMeta = util.meta.get("timer")

local function getTimer()
    local t = {}
    setmetatable(t, timerMeta)
    return t
end

local function scheduleTimer(toSchedule)
    local scheduledTime = toSchedule.scheduledTime

    for i=#scheduledTimers, 1, -1 do
        if scheduledTimers[i].scheduledTime > scheduledTime then
            table.insert(scheduledTimers, i+1, toSchedule)
            return
        end
    end

    table.insert(scheduledTimers, 1, toSchedule)
end

function timerMeta.funcs:cancel()
    for i=1, #scheduledTimers do
        if scheduledTimers[i] == self then
            table.remove(scheduledTimers, i)
        end
    end
end

function timer.simple(timeout, callback)
    local newTimer = getTimer()
    newTimer.scheduledTime = game.getElapsedTime() + timeout
    newTimer.callback = ceres.wrapCatch(callback)

    scheduleTimer(newTimer)
    return newTimer
end

function timer.periodic(period, callback)
    local newTimer = getTimer()
    newTimer.scheduledTime = game.getElapsedTime() + period
    newTimer.period = period
    newTimer.callback = ceres.wrapCatch(callback)

    scheduleTimer(newTimer)
    return newTimer
end

hook.once(function()
    local ti = CreateTimer()
    TimerStart(ti, 1/128, true, function()
        local elapsed = game.getElapsedTime()
        local toReschedule = {}

        for i=#scheduledTimers, 1, -1 do
            local currentTimer = scheduledTimers[i]
            if currentTimer.scheduledTime <= elapsed then
                scheduledTimers[i] = nil
                currentTimer.callback()

                if currentTimer.period then
                    currentTimer.scheduledTime = elapsed + currentTimer.period
                    table.insert(toReschedule, currentTimer)
                end
            else
                break
            end
        end

        for _, v in pairs(toReschedule) do
            scheduleTimer(v)
        end
    end)
end)

File: ts/src/tsstd/core/timer.ts
import { elapsedTime } from "./game"

class Timer {
    private scheduledTime: number
    private period?: number
    private callback: (this: void) => void

    private static context = util.contextFn("timer", () => {
        Timer.initialize()

        return {
            scheduled: [] as Timer[]
        }
    })

    private constructor(scheduledTime: number, callback: () => void) {
        this.scheduledTime = scheduledTime
        this.callback = callback
    }

    cancel() {
        let scheduled = Timer.context.scheduled
        for (let i = 0; i < scheduled.length; i++) {
            if (scheduled[i] == this) {
                // performance workaround...
                table.remove(scheduled, i + 1)
            }
        }
    }

    static simple(timeout: number, callback: () => void): Timer {
        let timer = new Timer(elapsedTime() + timeout, ceres.wrapSafeCall(callback))
        timer.schedule()
        return timer
    }

    static periodic(period: number, callback: () => void): Timer {
        let timer = new Timer(elapsedTime() + period, ceres.wrapSafeCall(callback))
        timer.period = period
        timer.schedule()
        return timer
    }

    private schedule() {
        let scheduledTime = this.scheduledTime
        let scheduled = Timer.context.scheduled

        for (let i = scheduled.length - 1; i >= 0; i--) {
            if (scheduled[i].scheduledTime > scheduledTime) {
                // apparently, there's no efficient way of doing this in TS
                // without using splice...
                table.insert(scheduled, i + 1 + 1, this)
                return
            }
        }

        table.insert(scheduled, 1, this)
    }

    private static initialize() {
        let t = CreateTimer()
        TimerStart(t, 1 / 128, true, () => {
            let elapsed = elapsedTime()
            let toReschedule: Timer[] = []
            let scheduled = Timer.context.scheduled

            for (let i = scheduled.length - 1; i >= 0; i--) {
                let currentTimer = scheduled[i]
                if (currentTimer.scheduledTime <= elapsed) {
                    delete scheduled[i]
                    currentTimer.callback()

                    if (currentTimer.period != undefined) {
                        currentTimer.scheduledTime = elapsed + currentTimer.period
                        toReschedule.push(currentTimer)
                    }
                } else {
                    break
                }
            }

            for (let v of toReschedule) {
                v.schedule()
            }
        })
    }
}

export default Timer
